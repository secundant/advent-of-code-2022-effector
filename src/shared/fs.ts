import { readFile } from 'node:fs/promises';

export const readRelativeFile = (meta: ImportMeta, path: string) =>
  readFile(new URL(path, meta.url).pathname, 'utf-8');
