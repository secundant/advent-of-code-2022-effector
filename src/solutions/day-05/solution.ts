import { combine, createStore } from 'effector';

export interface MoveCommand {
  size: number;
  from: number;
  to: number;
}

const move = (stacks: string[][], { to, size, from }: MoveCommand) => {
  stacks[to].push(...stacks[from].splice(stacks[from].length - size).reverse());
};

export const $stacks = createStore<string[][]>([], { sid: '$stacks' });
export const $commands = createStore<MoveCommand[]>([], { sid: '$commands' });

export const $reorderedStacks = combine($stacks, $commands, (stacks, commands) =>
  commands.reduce(
    (acc, command) => {
      move(acc, command);
      return acc;
    },
    Array.from(stacks, stack => Array.from(stack))
  )
);

export const $reorderedTopCrates = $reorderedStacks.map(stacks =>
  stacks.map(stack => stack.at(-1))
);
export const $reorderedMask = $reorderedTopCrates.map(chars =>
  chars
    .filter(Boolean)
    .map(crateName => crateName!.slice(1, 2))
    .join('')
);
