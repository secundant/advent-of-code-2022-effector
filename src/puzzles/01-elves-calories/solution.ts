import { createEvent, createStore, sample } from 'effector';
import { max, sum } from '@/shared/math';

export const elvesWroteNewStatistics = createEvent<string>();

const $detailedCalories = createStore<number[][]>([]);

export const $maxCalories = $detailedCalories.map(stats => max(stats.map(sum)));

sample({
  clock: elvesWroteNewStatistics.map(text =>
    text
      .trim()
      .split('\n\n')
      .map(chunk => chunk.split('\n').map(line => Number.parseInt(line, 10)))
  ),
  target: $detailedCalories
});
