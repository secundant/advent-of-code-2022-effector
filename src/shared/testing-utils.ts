import { readFile } from 'node:fs/promises';

export const createTestContext = (meta: ImportMeta) => {
  const loadFile = (name = './data.txt') => readFile(new URL(name, meta.url).pathname, 'utf-8');

  return {
    loadFile
  };
};
