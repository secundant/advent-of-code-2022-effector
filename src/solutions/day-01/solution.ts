import { createEvent, createStore, sample } from 'effector';
import { sum } from '@/shared/math';
import { last, numericComparator, sort } from '@/shared/std';

const countOfElvesCarryingTheMostCalories = 3;

export const replaceCalories = createEvent<number[][]>();

const $rawCaloriesGroups = createStore<number[][]>([]);

const $calories = $rawCaloriesGroups.map(groups => groups.map(sum));
const $sortedCalories = $calories.map(calories => sort(calories, numericComparator));

export const $sumOfTopCalories = $sortedCalories.map(calories =>
  sum(calories.slice(-1 * countOfElvesCarryingTheMostCalories))
);
export const $maxCalories = $sortedCalories.map(last);

sample({
  clock: replaceCalories,
  target: $rawCaloriesGroups
});
