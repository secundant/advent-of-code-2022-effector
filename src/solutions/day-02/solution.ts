import { createStore } from 'effector';
import { sum } from '@/shared/math';

type Outcome = keyof typeof outcomeScore;
type GameShape = keyof typeof shapes;
type GameRound = [GameShape, GameShape];
type Encrypted = [keyof typeof reqCode, keyof typeof resCode];

const shapes = {
  rock: 'scissors',
  paper: 'rock',
  scissors: 'paper'
} as const;

const shapeScore = { rock: 1, paper: 2, scissors: 3 };
const outcomeScore = { lose: 0, draw: 3, win: 6 } satisfies Record<string, number>;

const reqCode = { A: 'rock', B: 'paper', C: 'scissors' } satisfies Record<string, GameShape>;
const resCode = { X: 'rock', Y: 'paper', Z: 'scissors' } satisfies Record<string, GameShape>;

const decrypt = ([req, res]: Encrypted): GameRound => [reqCode[req], resCode[res]];
const isWon = ([req, res]: GameRound) => shapes[res] === req;
const getRoundOutcome = ([req, res]: GameRound): Outcome =>
  isWon([req, res]) ? 'win' : isWon([res, req]) ? 'lose' : 'draw';
const getRoundTotalScore = ([req, res]: GameRound) =>
  outcomeScore[getRoundOutcome([req, res])] + shapeScore[res];

export const $encryptedRounds = createStore<Encrypted[]>([], { sid: '$encryptedRounds' });
export const $rounds = $encryptedRounds.map(encrypted => encrypted.map(decrypt));
export const $roundsScore = $rounds.map(rounds => rounds.map(getRoundTotalScore));
export const $totalScore = $roundsScore.map(sum);
