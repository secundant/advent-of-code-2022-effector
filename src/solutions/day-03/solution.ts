import { createStore } from 'effector';
import { sum } from '@/shared/math';
import { divideString, getCharCode, isUpperCase, uniq } from '@/shared/std';

export const compartmentsCount = 2;

export const getPriority = (char: string) =>
  isUpperCase(char)
    ? 27 + getCharCode(char) - getCharCode('A')
    : 1 + getCharCode(char) - getCharCode('a');

export const $encryptedRucksacks = createStore<string[]>([], { sid: '$encryptedRucksacks' });

export const $rucksacks = $encryptedRucksacks.map(encrypted =>
  encrypted.map(entry =>
    divideString(entry, entry.length / compartmentsCount).map(part => Array.from(part))
  )
);

export const $rucksacksErrors = $rucksacks.map(rucksacks =>
  rucksacks
    .map(([firstCompartment, ...otherCompartments]) =>
      firstCompartment.filter(itemCode =>
        otherCompartments.some(compartment => compartment.includes(itemCode))
      )
    )
    .map(uniq)
);

export const $failedItemsTotalPriority = $rucksacksErrors
  .map(rucksacks => rucksacks.map(rucksack => rucksack.map(getPriority)).map(sum))
  .map(sum);
