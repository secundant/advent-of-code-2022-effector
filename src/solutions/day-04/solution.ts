import { createStore } from 'effector';
import { toInt } from '@/shared/std';

type SerializedPair = [string, string];
type Range = [number, number];

const parseRange = (input: string) => input.split('-').map(toInt) as [number, number];
const parsePair = (pair: SerializedPair) => pair.map(parseRange) as [Range, Range];

const inRange = (value: number, [from, to]: Range) => value >= from && value <= to;
const isSubRange = (range: Range, subRange: Range) =>
  subRange.every(value => inRange(value, range));

const isOverlappingRange = (range: Range, subRange: Range) =>
  subRange.some(value => inRange(value, range));

const isFullyOverlappingPair = ([left, right]: [Range, Range]) =>
  isSubRange(left, right) || isSubRange(right, left);

const isPartiallyOverlappedPair = ([left, right]: [Range, Range]) =>
  isOverlappingRange(left, right) || isOverlappingRange(right, left);

export const $serializedPairs = createStore<SerializedPair[]>([], { sid: '$serializedPairs' });

export const $pairs = $serializedPairs.map(pairs => pairs.map(parsePair));

export const $fullyOverlappedPairs = $pairs.map(pairs => pairs.filter(isFullyOverlappingPair));

export const $anyOverlappedPairs = $pairs.map(pairs => pairs.filter(isPartiallyOverlappedPair));
