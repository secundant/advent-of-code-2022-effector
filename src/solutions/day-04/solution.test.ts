import { fork } from 'effector';
import { describe, expect, test } from 'vitest';
import { createTestContext } from '@/shared/testing-utils';
import { $fullyOverlappedPairs, $serializedPairs } from '@/solutions/day-04/solution';

const exampleData = `
2-4,6-8
2-3,4-5
5-7,7-9
2-8,3-7
6-6,4-6
2-6,4-8
`;

describe('Day 04', async () => {
  const ctx = createTestContext(import.meta);
  const createScope = (data: string) =>
    fork({
      values: [
        [
          $serializedPairs,
          data
            .trim()
            .split('\n')
            .map(line => line.split(','))
        ]
      ]
    });

  test('01 - should pass example', () => {
    expect(createScope(exampleData).getState($fullyOverlappedPairs)).toHaveLength(2);
  });

  test('01 - In how many assignment pairs does one range fully contain the other?', async () => {
    expect(
      createScope(await ctx.loadFile()).getState($fullyOverlappedPairs).length
    ).toMatchSnapshot();
  });
});
