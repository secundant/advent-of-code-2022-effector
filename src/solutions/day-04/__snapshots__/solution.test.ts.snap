// Vitest Snapshot v1

exports[`Day 04 > 01 - In how many assignment pairs does one range fully contain the other? 1`] = `459`;

exports[`Day 04 > 02 - In how many assignment pairs do the ranges overlap? 1`] = `779`;
